common/src/main/java/org/conscrypt/Hpke.java

```diff
@@ -0,0 +1,236 @@
+import java.security.PrivateKey;
+import java.security.PublicKey;
+
+/**
+ * Hybrid Public Key Encryption (HPKE) APIs.
+ * <p>
+ * HPKE allows multiple cryptographic operations to be done based on a given setup transaction.
+ * Therefore, the following APIs are stateful after setting up the transaction:
+ * <li>
+ *     <ul>{@link Hpke#open(byte[], byte[])} </ul>
+ *     <ul>{@link Hpke#seal(byte[], byte[])}</ul>
+ *     <ul>{@link Hpke#export(int, byte[])} </ul>
+ * </li>
+ *
+ * @see <a href="https://www.rfc-editor.org/rfc/rfc9180.html#hpke-export">HPKE RFC 9180</a>
+ */
+public class Hpke {
+  /** Represents the previous setup transaction mode. */
+  private enum Mode {
+    NONE,
+    SENDER,
+    RECIPIENT
+  }
+
+  private final HpkeSuite mHpkeSuite;
+
+  private HpkeContextHelper mContextHelper;
+  private EVP_HPKE_CTX mCtx;
+  private byte[] mEnc;
+
+  private Mode mSetupMode;
+
+  /**
+   * Constructor defining the HPKE scheme to be used. Current supported schemes are:
+   * <li>
+   *     <ul>{@link HpkeSuite#DHKEM_X25519_HKDF_SHA256_HKDF_SHA256_AES_128_GCM}</ul>
+   *     <ul>{@link HpkeSuite#DHKEM_X25519_HKDF_SHA256_HKDF_SHA256_AES_256_GCM}</ul>
+   *     <ul>{@link HpkeSuite#DHKEM_X25519_HKDF_SHA256_HKDF_SHA256_CHACHA20POLY1305}</ul>
+   * </li>
+   *
+   * @param hpkeSuite KEM, KDF, and AEAD scheme to be used
+   */
+  public Hpke(HpkeSuite hpkeSuite) {
+    mHpkeSuite = hpkeSuite;
+    mSetupMode = Mode.NONE;
+    mContextHelper = new HpkeContextHelper();
+  }
+
+  /**
+   *  Creates an instance of {@link Hpke} meant for testing only by supplying an additional
+   *  {@link HpkeContextHelper}.
+   */
+  @Internal
+  public static Hpke createForTestingOnly(HpkeSuite hpkeSuite, HpkeContextHelper contextHelper) {
+    final Hpke hpke = new Hpke(hpkeSuite);
+    hpke.mContextHelper = contextHelper;
+    return hpke;
+  }
+
+  /**
+   * Hybrid Public Key Encryption (HPKE) decryption.
+   * <p>
+   * Returns a plaintext in a byte array given a required ciphertext and an optional associated
+   * data (aad). The API has a pre-requirement, an API call to set up[Mode]Recipient is needed so
+   * the API could be properly initialized. Note: This API is stateful.
+   * <p><p>
+   * Multiple messages decryption:<p>
+   * If decrypting multiple messages that were encrypted using the same context in a sequence,
+   * they must be decrypted in the same order as well. The API setup[Mode]Recipient must be called
+   * just once before multiple calls to this API.
+   * <p><p>
+   * Single-shot decryption:<p>
+   * If decryption a single message, the API setup[Mode]Recipient must be called
+   * before every single message. Note: Calling setup[Mode][Sender|Recipient] resets the HPKE
+   * context.
+   *
+   * @param ciphertext contains the encrypted plaintext
+   * @param aad        optional associated data
+   * @return plaintext in a byte array.
+   * @throws IllegalStateException if the API setup[Mode]Recipient hasn't been called or if an
```

Comment:
Why the two-state API? That seems unnecessarily error-prone. Instead you could make `setup[Mode]Recipient` be static functions that return new `Hpke` objects.

Reply:
With the new changes, setup[Mode]Sender will return the `enc` as part of the response. As the API is already checking if the setup operation has been called, returning an `IllegalArgumentException` if not called correctly, I'm thinking the caller will be able to quickly identify this during integration.

Reply:
This still exposes this half-initialized state to the caller. See go/totw/42. It is about C++, but I think it applies equally to Java as well.  It additionally doesn't reflect the specification.

This is an API we'd have to keep stable indefinitely, so we should take the time to get it right. Modeling the true state of the object in the type system shifts these kinds of mistakes left, so that they're caught at compile time, or avoided altogether because the invariants are clear from the API.

Absent a clear reason (I asked _why_ the two-state API), we should align with best practices and the specification.

Reply:
I applied the suggestion, I think it looks so much better. 

While doing that I splitted HpkeContext into HpkeContextSender and HpkeContextRecipient so it is more user friendly into knowing which methods each participant can call. Calling the method `setup[Mode]` initializes them. 

Regarding the class `HpkeContextSender`, as the setup call also initializes the encapsulated key, I made this available through a `getEnc` method. As the setup method is a factory method, it makes more sense to just return the class instance. If the caller is interested, they can retrieve the encapsulated key by calling the method. I added this on the factory method documentation.

Reply:
Yup!